new.go

```diff
@@ -103,145 +104,166 @@ func newContainerIDMappingOptions(idmapOptions *IDMappingOptions) storage.IDMapp
+func resolveImage(ctx context.Context, systemContext *types.SystemContext, store storage.Store, options BuilderOptions) (types.ImageReference, string, *storage.Image, error) {
+	type dockerPullCandidate struct {
+		ref   types.ImageReference
+		alias *shortnames.Resolved
+	}
+	var pullCandidates []dockerPullCandidate
+
+	if systemContext == nil {
+		systemContext = &types.SystemContext{}
+	}
+	systemContext.ShortNameMode = getShortNameMode()
+
+	// If the image name includes a transport we can use it as it.  Special
+	// treatment for docker references which are subject to pull policies
+	// that we're handling below.
+	srcRef, err := alltransports.ParseImageName(options.FromImage)
+	if err == nil {
+		if srcRef.Transport().Name() == docker.Transport.Name() {
+			pullCandidates = append(pullCandidates, dockerPullCandidate{ref: srcRef})
+		} else {
 			pulledImg, pulledReference, err := pullAndFindImage(ctx, store, srcRef, options, systemContext)
-			if err != nil {
-				logrus.Debugf("unable to pull and read image %q: %v", image, err)
-				failures = append(failures, failure{resolvedImageName: image, err: err})
-				continue
-			}
-			return pulledReference, transport, pulledImg, nil
+			return pulledReference, srcRef.Transport().Name(), pulledImg, err
 		}
+	}
 
-		destImage, err := localImageNameForReference(ctx, store, srcRef)
-		if err != nil {
-			return nil, "", nil, errors.Wrapf(err, "error computing local image name for %q", transports.ImageName(srcRef))
+	localImageRef, _, localImage, err := resolveLocalImage(systemContext, store, options)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	// If we could resolve the image locally, check if it was referenced by
+	// ID.  In that case, we don't need to bother any further and can
+	// prevent prompting the user.
+	if localImage != nil && strings.HasPrefix(localImage.ID, options.FromImage) {
+		return localImageRef, localImageRef.Transport().Name(), localImage, nil
+	}
+
+	if options.PullPolicy == PullNever || options.PullPolicy == PullIfMissing {
+		if localImage != nil {
+			return localImageRef, localImageRef.Transport().Name(), localImage, nil
 		}
-		if destImage == "" {
-			return nil, "", nil, errors.Errorf("error computing local image name for %q", transports.ImageName(srcRef))
+		if options.PullPolicy == PullNever {
+			return nil, "", nil, errors.Errorf("pull policy is %q but %q could not be found locally", "never", options.FromImage)
 		}
-		ref, err := is.Transport.ParseStoreReference(store, destImage)
+	}
+
+	// No transport, so let's use short-name resolution to find potential
+	// pullCandidates.
+	if len(pullCandidates) == 0 {
+		candidates, err := shortnames.Resolve(systemContext, options.FromImage)
 		if err != nil {
-			return nil, "", nil, errors.Wrapf(err, "error parsing reference to image %q", destImage)
+			return nil, "", nil, err
+		}
+		for i, candidate := range candidates {
+			ref, err := docker.ParseReference("//" + candidate.String())
```

Comment:
Eventually this should be `docker.NewReference(candidate.Value)` — after `.Value` is always normalized, which is not currently the case.